001-config-overhaul.md

```diff
@@ -648,6 +688,138 @@ hiccups along the way.
+3. **Documentation**: This behaviour will be explicitly documented in the
+   [configuration reference](#proposal-2---reference-before-implementation) to ensure users are aware of how relative
+   paths are handled.
+
+Standardising how relative paths are resolved can help prevent potential issues and confusion for users.
+
+#### Summary
+
+This proposal seeks to bring consistency and clarity to how relative file paths are resolved in Iroha's configuration.
+Paths in the config file will be relative to its location, while paths from environment variables will be relative to
+the CWD.
+
+### Proposal 11 - Adopt Same Proposals for Iroha Client CLI
+
+#### Objective
+
+To apply the configuration improvement proposals devised for Iroha 2 to the `iroha_client_cli`.
+
+#### Rationale
+
+1. **Shared Foundation**: The iroha_client_cli and Iroha 2 are built on the same foundation, and the client CLI's
+   configuration structure is similar to Iroha's.
+2. **Consistent Experience**: Providing a consistent configuration experience between Iroha and its official client
+   ensures that users and administrators familiar with one will find the other intuitive.
+3. **Efficiency**: By applying the same changes and improvements to both, we can leverage the same tools, documentation
+   updates, and testing procedures, leading to efficient development and maintenance.
+4. **Prevent Duplication**: Ensuring both tools share the same configuration enhancements reduces the chance of
+   duplicating efforts or inconsistencies in future updates.
+
+#### Proposal Details
+
+All proposals and changes specified for Iroha 2's configuration will be mirrored in `iroha_client_cli`. This includes
+(but is not limited to):
+
+- Adopting the same naming conventions.
+- Applying the same handling for relative paths.
+- Implementing the same structure and organisation for configuration fields.
+- Utilising the same documentation standards and resources.
+
+#### Summary
+
+To ensure a consistent user experience and efficient development, the configuration improvements proposed for Iroha 2
+will also be implemented in the `iroha_client_cli`. This alignment will guarantee that users and developers encounter a
+uniform configuration environment across both tools.
+
+### Proposal 12 - Handling Configuration-ISI Discrepancies
+
+#### Objective
+
+Ensure that the new configuration system gracefully and reliably interacts with the existing on-chain configuration set
+by Iroha Special Instructions (ISI), even as the nature and implementation of these instructions are still evolving.
+
+#### Rationale
+
+The existing ISI mechanism allows certain configuration parameters to be set on-chain. These on-chain configurations, if
+not handled correctly, could lead to inconsistencies or unexpected behaviours in nodes that might have differing local
+configurations.
+
+1. **Clarify Hierarchies**: Establish a clear priority hierarchy between local configurations and on-chain
```

Comment:
would it make sense to move network-related properties to the on-chain exclusively?

Reply:
I think it would. We need other opinions here.

Reply:
Also, all options under `wsv` namespace fall under this category, imo. They configure limits of the entitires, e.g. `wsv.asset_metadata_limits`, and WASM Runtime limits.

Reply:
Oh, I discovered that WSV parameters are already configurable on-chain:

https://github.com/hyperledger/iroha/blob/fc744249c988e0dbc972c5bcdce07f18ff08731c/configs/peer/genesis.json#L120-L174

It makes even more sense to remove on-chain parameters from the local configuration at all.

Reply:
The only problem with that is the following scenario imo, imagine an empty (just having the genesis) peer gets registered in an already running network where such parameters got changed via committed transactions.

The new peer must sync. The main problem is outdated topology. What if all peers from initial `config` got replaced with others? In that case does it mean each peer should rewrite its `trusted_peers` (each time some peer registers/unregisters) and restart? The same is applicable to the new peer. It seems it should get the latest/relevant state of the config from somewhere ignoring the 'initial' topology.

As for the `wsv` parameters, I'm not sure if there is a risk of related failure since such parameter values are expected to increase only.

What about other parts of the config (not `wsv`)?

Reply:
> It makes even more sense to remove on-chain parameters from the local configuration at all.

Do you think we should have a separate config for on-chain parameters? `genesis.json` can reference this file or Iroha can just expand genesis block with a transaction containing on-chain parameters from this file

Reply:
@Mingela, I cannot fully comprehend the problem of `trusted_peers` you described. `sumeragi.trusted_peers` is currently configurable only through the config file, and is only applied on start-up. The only way to update topology is to use [`Register<Peer>` ISI](https://hyperledger.github.io/iroha-2-docs/guide/blockchain/instructions.html#un-register). Aaaand yeah, I think it is out of scope of this RFC. It belongs to more major Iroha problems: genesis & consensus mechanisms.

> As for the `wsv` parameters, I'm not sure if there is a risk of related failure since such parameter values are expected to increase only.

`wsv` parameters configure limits for entities & smartcontracts execution. If several peers have different limits, and one rejects txs accepted by another, there will be no consensus. Therefore, `wsv` parameters directly affect consensus between peers, and that is why I think they should be configured on-chain exclusively.

> What about other parts of the config (not `wsv`)?

You might check the [draft of the new config reference](https://iroha-2-docs-git-fork-0x009922-392-draft-7a77f6-irohas-projects.vercel.app/api/config/). It is raw, but you can see explore the parameters. Shortly:

- **Base parameters:** pub key, priv key, and p2p addr of the peer - they are all local
- **Genesis account keys:** should be set locally for all peers (priv key only for the submitting one)
- **Sumeragi:** most parameters makes sense to make on-chain - block/commit time, blocks/txs gossiping. Only `sumeragi.trusted_peers`, I guess, is local for a peer.
- **Torii:** configures peer's API gateway, everything is mostly local - addr, transaction endpoint max-content, and lazy-query behaviour
- **Queue:** queue of transactions to make blocks from. I am not sure here, e.g. `queue.transaction_time_to_live` might be on-chain, while `queue.future_threshold` is more a local one.
- **Kura:** purely local, configures how particular peer works with the storage
- **Logger:** purely local
- **Telemetry:** purely local
- **Snapshot:** purely local

@mversic,

> Do you think we should have a separate config for on-chain parameters? `genesis.json` can reference this file or Iroha can just expand genesis block with a transaction containing on-chain parameters from this file

I don't know. If parameters are on-chain, then they should be configured with ISIs anyway. If these parameters might be configured on-chain exclusively, without allowing local overrides, then ISIs should be the only way, imo.

Anyway, I have a more general question: is on-chain configuration needed **now**?

Iroha 2.0 will not be suitable for public blockchain cases. The participating peers are generally known, and operators have a high level of coordination. We can remain only local configuration, which will make the system way more simple. We can provide clear configuration management guideline, which will ensure that while configurations might differ, they don't diverge too drastically. And we should provide tools so that operators could compare their settings with others or against recommended settings (in this case, `GET /configuration` will remain, but with chain-wide related fields only and in TOML format). This will be useful for troubleshooting and performance tuning.

This way has its own complications, but not that much as chain-wide configuration has. Ideally, the latter one should be accompanied by a mechanism to ensure that peers indeed behave accordingly to the config:

- Nodes should audit other nodes' behaviour and penalise them if their actual behaviour diverge from the configuration
- Operators should have monitoring tools to see if some nodes do not conform with chain-wide config

Also, while some parameters are "100% chain-wide", I couldn't label some other parameters that certainly.

Reply:
After discussion with the team, we decided that the best solution would be to remove chain-wide configuration from local file. I updated the RFC accordingly.